src/std_extensions/arithmetic/int_types.rs

```diff
@@ -1,127 +1,132 @@
 //! Basic integer types
 
+use std::num::NonZeroU64;
+
 use smol_str::SmolStr;
 
 use crate::{
-    extension::SignatureError,
     types::{
         type_param::{TypeArg, TypeArgError, TypeParam},
         ConstTypeError, CustomCheckFailure, CustomType, Type, TypeBound,
     },
     values::CustomConst,
     Extension,
 };
-
+use lazy_static::lazy_static;
 /// The extension identifier.
 pub const EXTENSION_ID: SmolStr = SmolStr::new_inline("arithmetic.int.types");
 
-/// Identfier for the integer type.
+/// Identifier for the integer type.
 const INT_TYPE_ID: SmolStr = SmolStr::new_inline("int");
 
-fn int_custom_type(n: u8) -> CustomType {
-    CustomType::new(
-        INT_TYPE_ID,
-        [TypeArg::USize(n as u64)],
-        EXTENSION_ID,
-        TypeBound::Copyable,
-    )
+fn int_custom_type(width_arg: TypeArg) -> CustomType {
+    CustomType::new(INT_TYPE_ID, [width_arg], EXTENSION_ID, TypeBound::Copyable)
 }
 
-/// Integer type of a given bit width.
+/// Integer type of a given bit width (specified by the TypeArg).
 /// Depending on the operation, the semantic interpretation may be unsigned integer, signed integer
 /// or bit string.
-pub fn int_type(n: u8) -> Type {
-    Type::new_extension(int_custom_type(n))
+pub(super) fn int_type(width_arg: TypeArg) -> Type {
+    Type::new_extension(int_custom_type(width_arg))
+}
+
+lazy_static! {
+    /// Array of valid integer types, indexed by log width of the integer.
+    pub static ref INT_TYPES: [Type; (MAX_LOG_WIDTH + 1) as usize] = (0..MAX_LOG_WIDTH + 1)
+        .map(|i| int_type(TypeArg::BoundedNat(i as u64)))
+        .collect::<Vec<_>>()
+        .try_into()
+        .unwrap();
 }
 
-fn is_valid_width(n: u8) -> bool {
-    (n == 1)
-        || (n == 2)
-        || (n == 4)
-        || (n == 8)
-        || (n == 16)
-        || (n == 32)
-        || (n == 64)
-        || (n == 128)
+const fn is_valid_log_width(n: u8) -> bool {
+    n <= MAX_LOG_WIDTH
 }
 
-/// Get the bit width of the specified integer type, or error if the width is not supported.
-pub fn get_width(arg: &TypeArg) -> Result<u8, SignatureError> {
-    let n: u8 = match arg {
-        TypeArg::USize(n) => *n as u8,
-        _ => {
-            return Err(TypeArgError::TypeMismatch {
-                arg: arg.clone(),
-                param: TypeParam::USize,
-            }
-            .into());
-        }
-    };
-    if !is_valid_width(n) {
-        return Err(TypeArgError::InvalidValue(arg.clone()).into());
+/// The largest allowed log width.
+pub const MAX_LOG_WIDTH: u8 = 7;
+
+/// Type parameter for the log width of the integer.
+// SAFETY: unsafe block should be ok as the value is definitely not zero.
+pub const LOG_WIDTH_TYPE_PARAM: TypeParam =
+    TypeParam::bounded_nat(unsafe { NonZeroU64::new_unchecked(MAX_LOG_WIDTH as u64 + 1) });
```

Comment:
Would this be avoidable by using a macro definition for `MAX_LOG_WIDTH`, so that the compiler can see that it's a safe value? @aborgna-q ?

Reply:
The safe alternative is to use `new()` but that produces an option (None if you gave it zero). We could happily unwrap it...but unwrap is not const yet, in the works.

Reply:
our [miri](https://github.com/rust-lang/miri) CI check should check this is fine